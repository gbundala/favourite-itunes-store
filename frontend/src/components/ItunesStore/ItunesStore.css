.main-wrapper {
  padding: 24px;
}

.project-items-wrapper {
  display: flex;
  flex-wrap: wrap;
  gap: 16px;
}

.empty-array-note {
  color: rgb(180, 173, 173);
}
