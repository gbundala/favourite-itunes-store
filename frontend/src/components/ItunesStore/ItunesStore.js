import React, { useState } from "react";
import SearchCriteria from "../SearchCriteria/SearchCriteria";
import ItunesItem from "../ItunesItem/ItunesItem";
import "./ItunesStore.css";

/**
 *
 * The WebProjects component is the main component which
 * handles all the major business logic of this application
 * of making the GET, POST, PUT and DELETE calls to the
 * endpoints exposed by the Express application using the fetch
 * method.
 *
 * The presentational bits of this app have been delegated over
 * to the respective children of this component allowing
 * this component to focus on the logic and data fetching
 * tasks. All event handlers are defined here as well as the
 * main state variables that are relevant to the most part
 * of the application.
 *
 * The useEffect has been used to make the GET request upon
 * mounting of the component on the DOM. Whereas the event
 * handlers has been used to make the POST, PUT and DELETE
 * request. The PUT and POST method includes headers and body
 * to pass data from the React app to the Express server
 */

export default function ItunesStore({
  favouriteItunes,
  setFavouriteItunes,
  itunesItems,
  setITunesItems,
}) {
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(null);

  function handleSearchItunes(searchCriteria) {
    const { searchTerm, mediaType } = searchCriteria;

    fetch(`/api/?term=${searchTerm}&media=${mediaType}`)
      .then((res) => res.json())
      .then(
        (data) => {
          setITunesItems(data.itunesItemsArrayWithIds);
          setLoading(false);
          setError(null);
        },
        (err) => {
          setError(err);
          setLoading(false);
          console.error("Error in fetching data: ", err);
        }
      );
  }

  function handleAddToFavouriteItunes(itunesItemAdded) {
    // FIXME: Delete consolelog
    console.log("handled", itunesItemAdded);
    const isItemAlreadyAdded = favouriteItunes.find((item) => {
      return item.uniqueId === itunesItemAdded.uniqueId;
    });

    if (!isItemAlreadyAdded) {
      // FIXME: Delete consolelog
      // TODO: Also find a way so that the ItunesItem component does not clear its state when returning back to the home page
      console.log("added?", isItemAlreadyAdded);
      setFavouriteItunes([...favouriteItunes, itunesItemAdded]);
    } else {
      alert("The item is already in favourites");
    }
  }

  return (
    <div className="main-wrapper">
      <h1>iTunes Store Items</h1>

      <SearchCriteria
        setLoading={setLoading}
        handleSearchItunes={handleSearchItunes}
      />

      {loading && <p>Loading...</p>}

      {error && <p>Something is wrong!</p>}

      <div className="project-items-wrapper">
        {itunesItems &&
          itunesItems.map((itunesItem) => {
            return (
              <ItunesItem
                key={itunesItem.uniqueId}
                itunesItem={itunesItem}
                handleAddToFavouriteItunes={handleAddToFavouriteItunes}
                isFromStore
              />
            );
          })}
      </div>
    </div>
  );
}
