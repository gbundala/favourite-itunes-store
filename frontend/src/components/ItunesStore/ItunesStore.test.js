// Import libraries and frameworks
import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import testRenderer from "react-test-renderer";

// Import Component to be tested
import ItunesStore from "./ItunesStore";

// FIXME: Check at the relevance of setupTest.js file

/**
 *
 * TODO: Write this top level description
 * This test cut across even to the child component ItunesItems as it tests that the fetch data is appropriately rendered on the page
 */

let targetContainerEl = null;
const fakeData = [
  {
    wrapperType: "track",
    kind: "song",
    artistId: 129335935,
    collectionId: 647928068,
    trackId: 647928086,
    artistName: "Wale",
    collectionName: "The Gifted",
    trackName: "Bad (Remix) [feat. Rihanna]",
    collectionCensoredName: "The Gifted",
    trackCensoredName: "Bad (Remix) [feat. Rihanna]",
    artistViewUrl: "https://music.apple.com/us/artist/wale/129335935?uo=4",
    collectionViewUrl:
      "https://music.apple.com/us/album/bad-remix-feat-rihanna/647928068?i=647928086&uo=4",
    trackViewUrl:
      "https://music.apple.com/us/album/bad-remix-feat-rihanna/647928068?i=647928086&uo=4",
    previewUrl:
      "https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview125/v4/79/ec/24/79ec241b-95c5-d746-8252-5d554e9be008/mzaf_1312633516429125863.plus.aac.p.m4a",
    artworkUrl30:
      "https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/2b/a9/e0/2ba9e01e-ceaf-adab-e95f-a6661656b2a7/source/30x30bb.jpg",
    artworkUrl60:
      "https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/2b/a9/e0/2ba9e01e-ceaf-adab-e95f-a6661656b2a7/source/60x60bb.jpg",
    artworkUrl100:
      "https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/2b/a9/e0/2ba9e01e-ceaf-adab-e95f-a6661656b2a7/source/100x100bb.jpg",
    collectionPrice: 9.99,
    trackPrice: 1.29,
    releaseDate: "2013-06-03T07:00:00Z",
    collectionExplicitness: "explicit",
    trackExplicitness: "explicit",
    discCount: 1,
    discNumber: 1,
    trackCount: 16,
    trackNumber: 10,
    trackTimeMillis: 238827,
    country: "USA",
    currency: "USD",
    primaryGenreName: "Hip-Hop/Rap",
    contentAdvisoryRating: "Explicit",
    isStreamable: true,
    uniqueId: 748860378379,
  },
];

// Setting up and Tearing down the the target element
// in order to isolate the effects of the test to itself

// SetUp
beforeEach(() => {
  // Setting up the targetElement for rendering during the tests
  // Then appending it to the body of the HTML
  targetContainerEl = document.createElement("div");
  document.body.appendChild(targetContainerEl);
});

// Tear Down
afterEach(() => {
  // Cleaning up the effects of the test
  unmountComponentAtNode(targetContainerEl);
  targetContainerEl.remove();
  targetContainerEl = null;
});

// UNIT Testing

test("Component Successfully fetches data and correctly renders", async () => {
  const fakeItunesData = fakeData;

  jest.spyOn(global, "fetch").mockImplementation(() =>
    Promise.resolve({
      json: () => Promise.resolve(fakeItunesData),
    })
  );

  await act(async () => {
    render(<ItunesStore itunesItems={fakeItunesData} />, targetContainerEl);
  });

  //   Assertions

  // Card title assertion
  expect(
    targetContainerEl.getElementsByClassName("card-title h5")[0].textContent
  ).toBe("The Gifted");

  //   Card sub-title assertion
  expect(
    targetContainerEl.getElementsByClassName(
      "mb-2 text-muted card-subtitle h6"
    )[0].textContent
  ).toBe("Wale");

  //   Removing the mock to isolate the test from
  //   other tests
  global.fetch.mockRestore();
});

// SNAPSHOT TESTING

// Testing whether the component renders correctly as intended
// inline with the saved snapshot

// With no data
test("ItunesStore component renders inline with snapshot", () => {
  // Provide the value of itunesStore to be
  // an array in order to avoid the bug when
  // itunesStore.length code is run
  const tree = testRenderer.create(<ItunesStore itunesItems={[]} />).toJSON();

  expect(tree).toMatchSnapshot();
});

// With data
test("ItunesStore component renders inline with snapshot with data", () => {
  const tree = testRenderer
    .create(<ItunesStore itunesItems={fakeData} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
