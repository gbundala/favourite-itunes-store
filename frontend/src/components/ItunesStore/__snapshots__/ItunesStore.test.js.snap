// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItunesStore component renders inline with snapshot 1`] = `
<div
  className="main-wrapper"
>
  <h1>
    iTunes Store Items
  </h1>
  <div
    className="add-item-wrapper"
  >
    <h4
      className="add-item-title"
    >
      Search iTunes and Apple Books Store
    </h4>
    <div
      className="mb-3 input-group"
    >
      <input
        className="form-control"
        name="searchTerm"
        onChange={[Function]}
        placeholder="Enter Search Term"
        value=""
      />
      <input
        className="form-control"
        name="mediaType"
        onChange={[Function]}
        placeholder="Enter Media Type"
        value=""
      />
      <button
        className="btn btn-outline-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Search iTunes
      </button>
    </div>
  </div>
  <div
    className="project-items-wrapper"
  />
</div>
`;

exports[`ItunesStore component renders inline with snapshot with data 1`] = `
<div
  className="main-wrapper"
>
  <h1>
    iTunes Store Items
  </h1>
  <div
    className="add-item-wrapper"
  >
    <h4
      className="add-item-title"
    >
      Search iTunes and Apple Books Store
    </h4>
    <div
      className="mb-3 input-group"
    >
      <input
        className="form-control"
        name="searchTerm"
        onChange={[Function]}
        placeholder="Enter Search Term"
        value=""
      />
      <input
        className="form-control"
        name="mediaType"
        onChange={[Function]}
        placeholder="Enter Media Type"
        value=""
      />
      <button
        className="btn btn-outline-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Search iTunes
      </button>
    </div>
  </div>
  <div
    className="project-items-wrapper"
  >
    <div>
      <link
        crossOrigin="anonymous"
        href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css"
        integrity="sha384-1BmE4kWBq78iYhFldvKuhfTAU6auU8tT94WrHftjDbrCEXSU1oBoqyl2QvZ6jIW3"
        rel="stylesheet"
      />
      <div
        className="container py-5 h-100 card-style card"
      >
        <img
          className="card-img-top"
          src="https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/2b/a9/e0/2ba9e01e-ceaf-adab-e95f-a6661656b2a7/source/100x100bb.jpg"
        />
        <div
          className="card-body"
        >
          <div
            className="card-title h5"
          >
            The Gifted
          </div>
          <div
            className="mb-2 text-muted card-subtitle h6"
          >
            Wale
          </div>
          <p
            className="card-text"
          />
          <div
            className="buttons-wrapper"
          >
            <button
              className="edit-button btn btn-light"
              data-testid="eventTrigger"
              disabled={false}
              id="favourite-button"
              onClick={[Function]}
              type="button"
            >
              Add to favourites
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
